hello-robot/stretch_web_teleop
# Generate IK URDF on-the-fly

# Description

Previously, Web Teleop's click-to-pregrasp feature required the user to run a script to generate a custom URDF for its inverse kinematics package. This step effectively cached a modified version of the calibrated URDF in `~/stretch_user`.

Caching URDFs is a bad idea because a robot's "calibrated" URDF is unique to it, and likely won't work as well or at all on another Stretch robot. Even within a single robot, calibration parameters (e.g. backlash, offset, etc.) change over time, and it's common for users to use Stretch Calibration to recalibrate their robot. By caching the IK urdf, Web Teleop isn't keeping up with the latest calibration. This can also make debugging complicated because developers can't trust if their application is using the latest calibration.

A solution is to generate the URDF your application needs on-the-fly. Everytime Web Teleop is launched, it takes the URDF in `~/stretch_user` and creates the IK urdf it needs. Recently, the [Stretch URDF](https://github.com/hello-robot/stretch_urdf/pull/27) package received a utility module to generate custom URDFs on-the-fly. This PR uses the new module to generate the IK URDFs it needs.

# Testing procedure

Create a [new ROS workspace](https://docs.hello-robot.com/0.3/installation/ros_workspace/). You'll want to modify the `.repos` file in Stretch Install to pull this branch for Web Teleop (i.e. replace `version: master` with `version: bugfix/dont_cache_urdf` on [line 33](https://github.com/hello-robot/stretch_install/blob/master/factory/22.04/stretch_ros2_humble.repos#L33)). Then launch the interface as normal. Ensure click to pregrasp works as usual.

# Before opening a pull request

From the top-level of this repository, run:


# To merge
